baseplate/lib/circuit_breaker/breaker.py

```diff
@@ -0,0 +1,127 @@
+from collections import deque
+from datetime import datetime
+from datetime import timedelta
+from enum import Enum
+from math import ceil
+from random import random
+from typing import Deque
+
+
+class BreakerState(Enum):
+    WORKING = "working"
+    TRIPPED = "tripped"
+    # trip immediately after failure
+    TESTING = "testing"
+
+
+class Breaker:
+    """Circuit breaker.
+
+    The circuit breaker has 3 states:
+    * WORKING (closed)
+    * TRIPPED (open)
+    * TESTING (half open)
+
+    During normal operation the circuit breaker is in the WORKING state.
+
+    When the number of failures exceeds the threshold the breaker moves to the TRIPPED state. It
+    stays in this state for the timeout period.
+
+    After the timeout period passes the breaker moves to the TESTING state. If the next attempt
+    is successful the breaker moves to the WORKING state. If the next attempt is a failure the
+    breaker moves back to the TRIPPED state.
+
+    :param name: full name/path of the circuit breaker
+    :param samples: number of previous results used to calculate the trip failure ratio
+    :param trip_failure_percent: the minimum ratio of sampled failed results to trip the breaker
+    :param trip_for: how long to remain tripped before resetting the breaker
+    :param fuzz_ratio: how much to randomly add/subtract to the trip_for time
+    """
+
+    _state: BreakerState = BreakerState.WORKING
+    _is_bucket_full: bool = False
+
+    def __init__(
+        self,
+        name: str,
+        samples: int = 20,
+        trip_failure_ratio: float = 0.5,
+        trip_for: timedelta = timedelta(minutes=1),
+        fuzz_ratio: float = 0.1,
+    ):
+        self.name = name
+        self.samples = samples
+        self.results_bucket: Deque = deque([], self.samples)
+        self.tripped_until: datetime = datetime.utcnow()
+        self.trip_threshold = ceil(trip_failure_ratio * samples)
+        self.trip_for = trip_for
+        self.fuzz_ratio = fuzz_ratio
+        self.reset()
+
+    @property
+    def state(self) -> BreakerState:
+        if self._state == BreakerState.TRIPPED and (datetime.utcnow() >= self.tripped_until):
+            self.set_state(BreakerState.TESTING)
+
+        return self._state
+
+    def register_attempt(self, success: bool) -> None:
+        """Register a success or failure.
+
+        This may cause the state to change.
+
+        :param success: Whether the attempt was a success (not a failure).
+        """
+        # This breaker has already tripped, so ignore the "late" registrations
+        if self.state == BreakerState.TRIPPED:
+            return
+
+        if not success:
+            self.failures += 1
+
+        if self._is_bucket_full and not self.results_bucket[0]:
```

Comment:
🔕 can we grab [this comment](https://github.com/reddit/baseplate.py/pull/565/files#diff-e79733da473c929b08e27ae53c7fe982df8bf0253a46c7e9540f2800db9ace27R83) or put something like it here?